library/src/main/java/org/apache/apex/malhar/lib/dedup/BoundedDedupOperator.java

```diff
@@ -0,0 +1,204 @@
+ * NOTE: Users can decide upon the proper value for this parameter by guessing the number of distinct keys 
+ * in the application. A appropriate value would be sqrt(num distinct keys). In case, the number of distinct keys is a
+ * huge number, leave it blank so that the default value of 46340 will be used. The rationale for using this number is
+ * that sqrt(max integer) = 46340. This implies that the number of buckets used will roughly be equal to the size of 
+ * each bucket, thus spreading the load equally among each bucket.
+ *
+ */
+@Evolving
+public class BoundedDedupOperator extends AbstractDeduper<Object>
+{
+  private static final long DEFAULT_CONSTANT_TIME = 0;
+  private static final int DEFAULT_NUM_BUCKETS = 46340;
+
+  // Required properties
+  @NotNull
+  private String keyExpression;
+
+  //Optional, but recommended to be provided by user
+  private int numBuckets = DEFAULT_NUM_BUCKETS;
+
+  private transient Class<?> pojoClass;
+
+  private transient Getter<Object, Object> keyGetter;
+
+  @InputPortFieldAnnotation(schemaRequired = true)
+  public final transient DefaultInputPort<Object> input = new DefaultInputPort<Object>()
+  {
+    @Override
+    public void setup(PortContext context)
+    {
+      pojoClass = context.getAttributes().get(PortContext.TUPLE_CLASS);
+    }
+
+    @Override
+    public void process(Object tuple)
+    {
+      processTuple(tuple);
+    }
+
+    @Override
+    public StreamCodec<Object> getStreamCodec()
+    {
+      return getDeduperStreamCodec();
+    }
+  };
+
+  public BoundedDedupOperator()
+  {
+    managedState = new ManagedTimeStateImpl();
+  }
+
+  @Override
+  public void setup(OperatorContext context)
+  {
+    if (numBuckets == 0) {
+      numBuckets = DEFAULT_NUM_BUCKETS;
+    }
+    ((ManagedTimeStateImpl)managedState).setNumBuckets(numBuckets);
+    TimeBucketAssigner timeBucketAssigner = new TimeBucketAssigner();
+    managedState.setTimeBucketAssigner(timeBucketAssigner);
+    super.setup(context);
+  }
+
+  @Override
+  public void activate(Context context)
+  {
+    keyGetter = PojoUtils.createGetter(pojoClass, keyExpression, Object.class);
+  }
+
+  @Override
+  public void deactivate()
+  {
+  }
+
+  @Override
+  protected long getTime(Object tuple)
+  {
+    return DEFAULT_CONSTANT_TIME;
+  }
+
+  @Override
+  protected Slice getKey(Object tuple)
+  {
+    Object key = keyGetter.get(tuple);
+    return new Slice(key.toString().getBytes());
+  }
+
+  protected StreamCodec<Object> getDeduperStreamCodec()
+  {
+    return new DeduperStreamCodec(keyExpression);
+  }
+
+  @Override
+  protected Future<Slice> getAsyncManagedState(Object tuple)
+  {
+    Future<Slice> valFuture = ((ManagedTimeStateImpl)managedState).getAsync(getBucketId(getKey(tuple)),
+        getKey(tuple));
+    return valFuture;
+  }
+
+  @Override
+  protected void putManagedState(Object tuple)
+  {
+    ((ManagedTimeStateImpl)managedState).put(getBucketId(getKey(tuple)), DEFAULT_CONSTANT_TIME,
```

Comment:
getKey can be called only once. No need to call it for both parameters of getAsync.
 Add a line note